File: src/tools/errorTracking/errorDetails.ts
import type { z } from "zod";
import type { Context, Tool } from "../types";
import { ErrorTrackingDetailsSchema } from "../../schema/tool-inputs";

const schema = ErrorTrackingDetailsSchema;

type Params = z.infer<typeof schema>;

export const errorDetailsHandler = async (context: Context, params: Params) => {
	const { issueId, dateFrom, dateTo } = params;
	const projectId = await context.getProjectId();

	const errorQuery = {
		kind: "ErrorTrackingQuery",
		dateRange: {
			date_from: dateFrom || new Date(Date.now() - 7 * 24 * 60 * 60 * 1000).toISOString(),
			date_to: dateTo || new Date().toISOString(),
		},
		volumeResolution: 0,
		issueId,
	};

	const errorsResult = await context.api.query({ projectId }).execute({ queryBody: errorQuery });
	if (!errorsResult.success) {
		throw new Error(`Failed to get error details: ${errorsResult.error.message}`);
	}

	return {
		content: [{ type: "text", text: JSON.stringify(errorsResult.data.results) }],
	};
};

const tool = (): Tool<typeof schema> => ({
	name: "error-details",
	description: `
        - Use this tool to get the details of an error in the project.
    `,
	schema,
	handler: errorDetailsHandler,
});

export default tool;

File: python/tools/error_tracking/error_details.py
import json
from datetime import datetime, timedelta

from api.client import is_error, is_success
from schema.tool_inputs import ErrorTrackingDetailsSchema
from tools.tool_definitions import get_tool_definition
from tools.types import Context, Tool, ToolResult


async def error_details_handler(context: Context, params: ErrorTrackingDetailsSchema) -> ToolResult:
    project_id = await context.get_project_id()

    date_from = params.dateFrom or datetime.now() - timedelta(days=7)
    date_to = params.dateTo or datetime.now()

    error_query = {
        "kind": "ErrorTrackingQuery",
        "dateRange": {"date_from": date_from.isoformat(), "date_to": date_to.isoformat()},
        "volumeResolution": 0,
        "issueId": str(params.issueId),
    }

    errors_result = await context.api.query(project_id).execute({"query": error_query})
    if is_error(errors_result):
        raise Exception(f"Failed to get error details: {errors_result.error}")

    assert is_success(errors_result)
    return ToolResult(content=json.dumps(errors_result.data.results))


def error_details_tool() -> Tool[ErrorTrackingDetailsSchema]:
    definition = get_tool_definition("error-details")

    return Tool(
        name="error-details",
        description=definition["description"],
        schema=ErrorTrackingDetailsSchema,
        handler=error_details_handler,
    )
